### PR TITLE
Toggle republication on/off for posts

### PR DESCRIPTION
**Pull request by joshdarby**
## Changes

This pull request makes the following changes:

- Adds a custom metabox to the post editor document sidebar to check/uncheck if the post should hide the Republication sharing widget
- Adds a new filter, `hide_republication_widget`, which allows developers to programmatically hide the sharing button on any specific post/page/taxonomy
   - It is currently set up to allow the filter to take priority. For example, if category A is hidden through the filter, but post 3 inside of category A wants to show the button, toggling the checkbox in the post editor won't make the button show.

Screen Shot 2019-08-15 at 4 39 56 PM

## Why

For #51

## Testing/Questions

Features that this PR affects:

- Republication sharing button on posts

Questions that need to be answered before merging:

- [x] @benlk Do you think the priority is ok as it is, with the filter taking priority? Or should it function a different way?

Steps to test this PR:

1. Make sure the on/off checkbox in the post editor works
2. Make sure the filter works as expected
3. Make sure there are no weird priority edge cases between the filter and on/off checkbox.

### CODE REVIEW

File: includes/class-widget.php [hunk 1]
```diff
@@ -38,6 +38,22 @@ public function widget( $args, $instance ) {
...
 
 		global $post;
+
+		$hide_republication_widget_on_post = get_post_meta( $post->ID, 'republication-tracker-tool-hide-widget' );
+
+		if( $hide_republication_widget_on_post ){
+			$hide_republication_widget_on_post = $hide_republication_widget_on_post[0];
+		}
+
+		$hide_republication_widget_filter = apply_filters( 'hide_republication_widget', false );
```
Comment (benlk) [inline includes/class-widget.php]: Why apply the filter to default-false when we could instead apply the filter to the value from the post meta? And should the filter pass the `global $post` as a second parameter on the filter, like in these examples?

https://github.com/INN/largo/blob/512da701664b329f2f92244bbe54880a6e146431/inc/ajax-functions.php#L203-L221

https://github.com/INN/super-cool-ad-inserter-plugin/blob/1da4ea864918c590a4bdb0afc7e1e0178c632b27/inc/scaip-shortcode-inserter.php#L130-L144

File: includes/class-article-settings.php [hunk 2]
```diff
@@ -103,6 +134,32 @@ public function render_metabox( $post, $args ) {
...
+
+			$checked = 'checked';
+
+		}
+
+		echo '<label>';
+			echo '<input type="checkbox" name="republication-tracker-tool-hide-widget" id="republication-tracker-tool-hide-widget" '.$checked.'>';
+			echo __('Hide the Republication sharing widget on this post?', 'republication-tracker-tool');
+		echo '</label>';
+
```
Comment (benlk) [inline includes/class-article-settings.php]: Should this output a warning if the value from the post meta does not match the value from the result of the filter for this post?

#51 describes:

> a comparison between the metadata value and the filter's value for the post that remarks whether the widget is programmatically excluded for the post.

File: docs/removing-republish-button-from-categories.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Removing the Republish Button From Specific Categories
+
+If you'd like to remove the Republish button from specific categories, the plugin includes a filter that will help you do this programatically. 
+
+The `hide_republication_widget` filter allows you to target specific posts/categories/tags/whatever you want to hide. All you have to do is check if the current post exists in the specific category, and tell the filter to return `true` if you'd like the button to be hidden.
+
+Here's an example of how to hide the Republish button on all posts in the category with an ID of 14.
+
+```
+function remove_republish_button_from_category(){
```
Comment (benlk) [inline docs/removing-republish-button-from-categories.md]: This example function needs to accept 1 parameter, and return the value of that parameter if the conditional is not set.

It should also have a docblock.

File: includes/class-article-settings.php [hunk 4]
```diff
@@ -60,6 +61,36 @@ public function register_meta_boxes() {
...
 		);
+
+		add_meta_box(
+			'republication-tracker-tool-hide-widget',
+			esc_html__( 'Hide Republication Widget', 'republication-tracker-tool' ),
+			array( $this, 'render_hide_widget_metabox' ),
+			array( 'post', 'page' ),
+			'side',
+			'default'
+		);
```
Comment (benlk) [inline includes/class-article-settings.php]: Should we set the seventh parameter of add_meta_box here to

```php
array(
    '__block_editor_compatible_meta_box' => true,
)
```

?

It's not explicitly required, but it may help forwards compatibility. https://developer.wordpress.org/block-editor/developers/backward-compatibility/meta-box/#testing-converting-and-maintaining-existing-meta-boxes

Do we want this checkbox's value to update on the gutenberg save action, rather than on page refresh?

File: includes/class-widget.php (hunk 1 above)
Comment (joshdarby) [reply includes/class-widget.php]: > Why apply the filter to default-false when we could instead apply the filter to the value from the post meta?

That's a good point I didn't think of. I'll update it.

Comment (joshdarby) [reply includes/class-widget.php]: > And should the filter pass the global $post as a second parameter on the filter, like in these examples?

@benlk What reason do you think we'd need `global $post` for in the filter?

File: includes/class-article-settings.php (hunk 4 above)
Comment (joshdarby) [reply includes/class-article-settings.php]: @benlk The new parameter is in https://github.com/INN/republication-tracker-tool/pull/54/commits/1002f263b842a31385d6723e6b7f9bb52182f7df.

File: includes/class-article-settings.php (hunk 2 above)
Comment (joshdarby) [reply includes/class-article-settings.php]: @benlk Instead of a warning, I updated the function so if the filter is currently being applied on the post you're editing then you'll see a message stating that instead of the option to check/uncheck it in https://github.com/INN/republication-tracker-tool/pull/54/commits/1002f263b842a31385d6723e6b7f9bb52182f7df.

File: includes/class-widget.php (hunk 1 above)
Comment (joshdarby) [reply includes/class-widget.php]: Updated to use the post meta as the default filter value in https://github.com/INN/republication-tracker-tool/pull/54/commits/c7889fbabe274437c517b292b18bf2bfebee5917.

File: docs/removing-republish-button-from-categories.md (hunk 3 above)
Comment (joshdarby) [reply docs/removing-republish-button-from-categories.md]: Docblock added in https://github.com/INN/republication-tracker-tool/pull/54/commits/17b1480cc99d5d5bd6c5626b6a2293261606d57a.

@benlk Can't we just `return false;` if the conditional isn't met?

File: includes/class-widget.php (hunk 1 above)
Comment (benlk) [reply includes/class-widget.php]: We should pass the current $post because filtering functions should have the post explicitly passed to them that they're operating on. I've been bitten too many times by situations where we expected global $post to be the correct post for the context that a function was running in, only to have global $post somehow not be correct. Not passing posts explicitly is a bad pattern.

File: docs/removing-republish-button-from-categories.md (hunk 3 above)
Comment (benlk) [reply docs/removing-republish-button-from-categories.md]: We cannot. There may be other filters here running at other priorities that set modify the value, or the post metabox may be set to `true`. Filters should not assume that the default value is set or unset; all that we can reasonably expect as the author of this filtering function is that the documentation regarding the type of the expected parameters is correct.

File: includes/class-article-settings.php [hunk 5]
```diff
@@ -103,6 +137,40 @@ public function render_metabox( $post, $args ) {
...
+		if( true == $hide_republication_widget ) {
+
+			$checked = 'checked';
+
+		}
+
+		$hide_republication_widget = apply_filters( 'hide_republication_widget', $hide_republication_widget );
+
+		if( true == $hide_republication_widget ){
+			echo '<p>The Republication sharing widget on this post is programatically disabled through the hide_republication_widget filter.</p>';
```
Comment (benlk) [inline includes/class-article-settings.php]: The only nitpicks I have here:

- the filter name should be wrapped in a code tag
- this should link to documentation somewhere

File: includes/class-article-settings.php (hunk 2 above)
Comment (benlk) [reply includes/class-article-settings.php]: 👍

File: includes/class-article-settings.php (hunk 5 above)
Comment (joshdarby) [reply includes/class-article-settings.php]: Updated in https://github.com/INN/republication-tracker-tool/pull/54/commits/8b3782d1090f9a4b8892f7cdd2f75a9156c44150. Note that the link won't work until this PR is merged.

File: includes/class-widget.php (hunk 1 above)
Comment (joshdarby) [reply includes/class-widget.php]: @benlk https://github.com/INN/republication-tracker-tool/pull/54/commits/fa3d283fa778bed9abf556a252558bb187f53b07 now passes the current `$post` into the filter.

File: docs/removing-republish-button-from-categories.md (hunk 3 above)
Comment (joshdarby) [reply docs/removing-republish-button-from-categories.md]: @benlk The doc function is updated in https://github.com/INN/republication-tracker-tool/pull/54/commits/45e4fd516691c883758fc70f3a70800235079f11.

Comment (joshdarby) [reply docs/removing-republish-button-from-categories.md]: @benlk https://github.com/INN/republication-tracker-tool/pull/54/commits/45e4fd516691c883758fc70f3a70800235079f11 updates the sample function to include the new params and returns the value of `$hide_republication_widget`.

Comment (benlk) [reply docs/removing-republish-button-from-categories.md]: 👍

File: includes/class-widget.php (hunk 1 above)
Comment (benlk) [reply includes/class-widget.php]: 👍

File: includes/class-article-settings.php (hunk 5 above)
Comment (benlk) [reply includes/class-article-settings.php]: Link not working until this PR is merged is fine.

Works with the current branch: https://github.com/INN/republication-tracker-tool/blob/51-toggle-republication-on-off-for-posts/docs/removing-republish-button-from-categories.md

**Comment by benlk** [comment]
Tested the disabling filter with:
```php
add_filter( 'hide_republication_widget', '__return_true', 10, 2 );
```

Screen Shot 2019-08-19 at 15 45 24

Works as expected, in the editor and the frontend.
